### ISSUE TITLE
SM120 Blackwell native self-test fails 128x64 / 64x64 sparse parity, causing Triton fallback

### ISSUE THREAD

**Issue by Tarve-t** [Zironic/H3-Optimizations#37]
### Short AI summary:

Current H3-Optimizations fails the native sparse self-test on my RTX PRO 6000 Blackwell SM120 because `128x64` and `64x64` parity return false, so it falls back to much slower Triton. The same native backend passes on my RTX 4090, and an older H3-Optimizations revision works normally on the 6000. Bypassing only that parity gate makes the native backend work again without obvious issues so far.

### System

* **GPU:** NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition
* **Compute capability:** SM120 / 12.0
* **Secondary GPU:** RTX 4090 / SM89
* **OS:** Windows 10
* **H3-Optimizations commit:** `d2fb37b`
* **Native backend:** ABI 3 / `native-v3`

### Problem

After updating H3-Optimizations to the current version, sparse attention started falling back to Triton:

```text
[WARNING] [H3 Optimizations] SPARSE ATTENTION FELL BACK to triton_sparse_int8.
INT8 Triton sparse is roughly half the speed of the native sparse kernel.
Reason: Kitchen INT8 unavailable: the installed comfy-kitchen has no block-sparse INT8 attention,
and the vendored library is unavailable: unknown;
Sparse Sage unavailable: Hybrid Sparse Attention requires the compiled spas_sage_attn package;
using INT8 Triton 64Q x 64KV sparse attention
```

The workflow itself was unchanged.

Updating the NVIDIA driver and rebooting did not change the behavior.

### Native DLL loads successfully

The included DLL exists at:

```text
native\bin\h3_int8_attention_v3.dll
```

and loads successfully:

```bat
python_embeded\python.exe -c "import torch,ctypes; ctypes.CDLL(r'ComfyUI\custom_nodes\H3-Optimizations\native\bin\h3_int8_attention_v3.dll'); print('DLL LOAD OK')"
```

Output:

```text
DLL LOAD OK
```

### RTX 4090 / SM89

Running native diagnostics on the RTX 4090 succeeds:

```text
GPU: NVIDIA GeForce RTX 4090 (8, 9)

native backend:
  ABI      : 3 (expected 3)
  build    : native-v3
  platform : windows-amd64
  route    : delta
  self-test: passed
```

### RTX PRO 6000 Blackwell / SM120

The same diagnostics on the RTX PRO 6000 give:

```text
GPU: NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition (12, 0)

native backend:
  ABI      : 3 (expected 3)
  build    : native-v3
  platform : windows-amd64
  route    : delta
  self-test: FAILED
```

Running the verbose self-test:

```bat
python_embeded\python.exe -c "import sys,torch; sys.path.insert(0,r'ComfyUI\custom_nodes\H3-Optimizations'); from h3_optimizations.native import selftest; print('GPU:',torch.cuda.get_device_name(1),torch.cuda.get_device_capability(1)); print(selftest.run(torch.device('cuda:1'),verbose=True))"
```

produces:

```text
GPU: NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition (12, 0)
  dense INT8 vs FP32 SDPA : rel_l2 0.014546 (tolerance 0.15) ok
  100% route vs dense     : 128x128=True, 128x64=False, 64x64=False
  sparse output + LSE     : max_abs 0.004799 ok

(False, {
    'int8_vs_sdpa_rel_l2': 0.014546,
    'full_route_bit_identical': {
        '128x128': True,
        '128x64': False,
        '64x64': False
    },
    'sparse_lse_output_bit_identical': True,
    'sparse_lse_max_abs': 0.004799
})
```

So the failure appears isolated to the sparse parity tests involving a 64-wide KV tile:

```text
128x128 = PASS
128x64  = FAIL
64x64   = FAIL
```

Dense INT8 accuracy and the sparse LSE test both pass.

### Regression test

I checked out the older revision:

```bat
git switch --detach f852ac8
```

With the same GPU, workflow, ComfyUI installation, and settings, the Triton fallback warning does not occur.

Returning to current `d2fb37b` reproduces the problem.

### Bypassing the parity gate

As a test, I changed the current self-test from:

```python
leg2 = all(parity.values())
```

to:

```python
leg2 = parity.get("128x128", False)
```

After deleting the cached self-test result and restarting ComfyUI, the current native backend initializes successfully:

```text
[INFO] [H3 Optimizations] armed: attention=sparse_kitchen_int8 ...
```

and H3 generation works normally so far.

I have not observed a CUDA crash or obvious visual corruption with the native backend enabled this way.

This makes me wonder whether the SM120 `128x64` / `64x64` kernels are producing a small numerical difference that fails the current bit-identical `torch.equal()` parity requirement, rather than the kernels being unusable outright.

Would it be possible to check the SM120 implementation for those two geometries, or whether the parity test should use a numerical tolerance on Blackwell instead of requiring bit-identical output?

**Comment by Zironic** [comment]
Newest commit should fix. Let me know if it does not.

**Comment by Tarve-t** [comment]
GPU: NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition (12, 0)
  dense INT8 vs FP32 SDPA : rel_l2 0.014817 (tolerance 0.15) ok
  100% route vs dense     : 128x128=True, 128x64=True, 64x64=True
  sparse output + LSE      : ok
(True, {'int8_vs_sdpa_rel_l2': 0.014817, 'full_route_bit_identical': {'128x128': True, '128x64': True, '64x64': True}, 'sparse_lse_output_bit_identical': True, 'sparse_lse_max_abs': 0.003024})

Seems to work. Thanks!

**Comment by Zironic** [comment]
duplicate of #36
